File: georocket-cli/src/test/java/io/georocket/commands/GetPropertyCommandTest.java
package io.georocket.commands;

import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;
import org.junit.Test;
import org.junit.runner.RunWith;

import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.get;
import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;

/**
 * Test for {@link GetPropertyCommand}
 * @author Tim Hellhake
 */
@RunWith(VertxUnitRunner.class)
public class GetPropertyCommandTest extends CommandTestBase<GetPropertyCommand> {
  @Override
  protected GetPropertyCommand createCommand() {
    return new GetPropertyCommand();
  }

  /**
   * Test a simple get
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void testGetProperty(TestContext context) throws Exception {
    String body = "a,b,c";
    String url = "/store/?property=test";
    stubFor(get(urlEqualTo(url))
      .willReturn(aResponse()
        .withStatus(200)
        .withBody(body)));

    Async async = context.async();
    cmd.setEndHandler(exitCode -> {
      context.assertEquals(0, exitCode);
      context.assertEquals(body, writer.toString());
      verifyRequested(url, context);
      async.complete();
    });

    cmd.run(new String[]{"-l", "/", "-prop", "test"}, in, out);
  }
}

File: georocket-cli/src/test/kotlin/io/georocket/commands/GetPropertyCommandTest.kt
package io.georocket.commands

import com.github.tomakehurst.wiremock.client.WireMock.aResponse
import com.github.tomakehurst.wiremock.client.WireMock.get
import com.github.tomakehurst.wiremock.client.WireMock.stubFor
import com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo
import io.vertx.ext.unit.TestContext
import io.vertx.ext.unit.junit.VertxUnitRunner
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Test for [GetPropertyCommand]
 */
@RunWith(VertxUnitRunner::class)
class GetPropertyCommandTest : CommandTestBase<GetPropertyCommand>() {
  override val cmd = GetPropertyCommand()

  /**
   * Test a simple get
   */
  @Test
  fun testGetProperty(context: TestContext) {
    val body = "a,b,c"
    val url = "/store/?property=test"
    stubFor(get(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(200)
            .withBody(body)))

    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf("-l", "/", "-prop", "test"), input, out)
      context.assertEquals(0, exitCode)
      context.assertEquals(body, writer.toString())
      verifyRequested(url, context)
      async.complete()
    }
  }
}
